### PR TITLE
[api][tools] Add swagger URL and UI for public APIs

### PR DESCRIPTION
**Pull request by tabraiz12**
1. API will be /api/swagger and /api/swagger
2. login can be accessed without signing in but would require authentication for checking any api
3. if logged in as an admin user can access any url

Add Swagger UI and URL for public url for easier QE Automation

Screenshot 2024-07-08 at 12 48 32 PM

## How was this patch tested?

- Manually Tested ui and json generation

Please review [Hue Contributing Guide](https://github.com/cloudera/hue/blob/master/CONTRIBUTING.md) before opening a pull request.

### CODE REVIEW

File: desktop/core/src/desktop/urls.py [hunk 1]
```diff
@@ -59,6 +63,20 @@
 handler500 = 'desktop.views.serve_500_error'
 
 
+schema_view = get_schema_view(
+   openapi.Info(
+      title="Snippets API",
```
Comment (tabraiz12) [inline desktop/core/src/desktop/urls.py:68]: will update title and details

File: desktop/core/base_requirements.txt [hunk 2]
```diff
@@ -73,6 +73,8 @@ git+https://github.com/gethue/django-babel.git
 django-utils-six==2.0
 six==1.16.0
 psutil==5.8.0
+drf-spectacular-sidecar==2024.7.1
+drf-spectacular==0.27.2
```
Comment (Harshg999) [inline desktop/core/base_requirements.txt:77]: Maybe `drf-spectacular[sidecar]==0.27.2` looks good?

File: desktop/core/src/desktop/urls.py [hunk 3]
```diff
@@ -202,6 +204,13 @@
 ]
 
 dynamic_patterns += [
+    re_path('api/schema',
+        SpectacularAPIView.as_view(permission_classes=(AllowAny,)),
+        name='schema'),
+  re_path('api/swagger',
+        SpectacularSwaggerView.as_view(url_name='schema', permission_classes=(AllowAny,)),
+        name='swagger-ui'),
+
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:213]: Missing redoc mapping?

File: desktop/core/src/desktop/urls.py [hunk 4]
```diff
@@ -202,6 +204,13 @@
 ]
 
 dynamic_patterns += [
+    re_path('api/schema',
+        SpectacularAPIView.as_view(permission_classes=(AllowAny,)),
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:208]: Why `permission_classes=(AllowAny,)`?

**Review by Harshg999 — COMMENTED** [review]
Nice, almost!
Also did you test it out E2E?

File: desktop/core/src/desktop/urls.py (hunk 3 above)
Comment (tabraiz12) [reply desktop/core/src/desktop/urls.py:213]: no intentionally removed it,  as we have swagger. we can add it if required.

File: desktop/core/src/desktop/urls.py (hunk 4 above)
Comment (tabraiz12) [reply desktop/core/src/desktop/urls.py:208]: removed it, this was to allow non logged in users to access the url, by default this is available. hence not required

**Comment by tabraiz12** [comment]
> Nice, almost! Also did you test it out E2E?

yes, i have tested this out.

File: desktop/core/base_requirements.txt [hunk 5]
```diff
@@ -73,6 +73,8 @@ git+https://github.com/gethue/django-babel.git
 django-utils-six==2.0
 six==1.16.0
 psutil==5.8.0
+drf-spectacular[sidecar]==0.27.2
+drf-spectacular==0.27.2
```
Comment (Harshg999) [inline desktop/core/base_requirements.txt:77]: To remove because above line is sufficient

File: desktop/core/src/desktop/urls.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 import sys
 import logging
 
+from drf_spectacular.views import SpectacularAPIView, SpectacularSwaggerView
+from rest_framework.permissions import AllowAny
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:25]: Remove unused import

File: desktop/core/src/desktop/urls.py [hunk 7]
```diff
@@ -202,6 +204,13 @@
 ]
 
 dynamic_patterns += [
+  re_path('api/schema',
+        SpectacularAPIView.as_view(),
+        name='schema'),
+  re_path('api/swagger',
+        SpectacularSwaggerView.as_view(url_name='schema'),
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:211]: We can remove `url_name='schema'` as well.

Also keep everything in one line per mapping?

File: desktop/core/src/desktop/urls.py [hunk 8]
```diff
@@ -202,6 +203,9 @@
 ]
 
 dynamic_patterns += [
+  re_path('api/schema', SpectacularAPIView.as_view(),name='schema'),
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:206]: We can remove this mapping as well, I checked and its just downloading the schema for all the APIs. Is this required?

File: desktop/core/src/desktop/urls.py [hunk 9]
```diff
@@ -202,6 +203,9 @@
...
+  re_path('api/swagger', SpectacularSwaggerView.as_view(),name='swagger-ui'),
```
Comment (Harshg999) [inline desktop/core/src/desktop/urls.py:207]: Let's change it to `'^api/swagger/?$'`?

File: desktop/core/src/desktop/settings.py [hunk 10]
```diff
@@ -325,8 +327,19 @@
...
+    'DEFAULT_AUTHENTICATION_CLASSES': desktop.conf.AUTH.API_AUTH.get(),
+    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
 }
+
+SPECTACULAR_SETTINGS = {
+    'TITLE': 'Hue Public API',
+    'VERSION': '1.0.0',
+    'SWAGGER_UI_DIST': 'SIDECAR',
+    'SWAGGER_UI_FAVICON_HREF': 'SIDECAR',
+    'REDOC_DIST': 'SIDECAR',
```
Comment (Harshg999) [inline desktop/core/src/desktop/settings.py:339]: Can remove if we're not adding redoc?

**Review by Harshg999 — COMMENTED** [review]
Few more review comments. Also please fix the linting issues flagged by CI checks.

File: desktop/core/src/desktop/urls.py (hunk 8 above)
Comment (tabraiz12) [reply desktop/core/src/desktop/urls.py:206]: yes, we require this for QE.

**Review by Harshg999 — APPROVED** [review]
LGTM! Thanks for addressing all the review comments.

File: desktop/core/src/desktop/urls.py (hunk 8 above)
Comment (Harshg999) [reply desktop/core/src/desktop/urls.py:206]: Okay then we can change the mapping here to '^api/schema/?$'?

**Review by JohanAhlen — APPROVED** [review]
LGTM!

**Pull request merged**